packages/block-library/src/table/edit.js

```diff
@@ -355,6 +361,30 @@ function TableEdit( {
 		}
 	}, [ hasTableCreated ] );
 
+	// We need to show the caption when changes come from
+	// history navigation(undo/redo).
+	useEffect( () => {
+		if ( caption && ! prevCaption ) {
+			setShowCaption( true );
+		}
+	}, [ caption, prevCaption ] );
+
+	useEffect( () => {
+		if ( ! isSelected && ! caption ) {
+			setShowCaption( false );
+		}
+	}, [ isSelected, caption ] );
```

Comment:
I'm not sure if syncing states via `useEffect` here is the best option. We might be able to achieve the same thing by deriving some of them here.

For instance, we could create a state `isCaptionToggled` and a derived state `showCaption` like this:

```js
const [ isCaptionToggled, setIsCaptionToggled ] = useState( !! caption );
const showCaption = !! caption || ( isCaptionToggled && isSelected );
```

Not sure if it will work though, needs more testing 😅 .